src/execution/join_hashtable.cpp

```diff
@@ -243,70 +441,249 @@ idx_t JoinHashTable::PrepareKeys(DataChunk &keys, vector<TupleDataVectorFormat>
+	idx_t remaining_count = count;
+
+	// use the ht bitmask to make the modulo operation faster but keep the salt bits intact
+	idx_t capacity_mask = ht->bitmask | aggr_ht_entry_t::SALT_MASK;
+
+	while (remaining_count > 0) {
+
+		idx_t salt_match_count = 0;
+		idx_t key_no_match_count = 0;
+
+		// iterate over each entry to find out whether it belongs to an existing list or will start
+		// a new list
+		for (idx_t i = 0; i < remaining_count; i++) {
+
+			const idx_t row_index = remaining_sel->get_index(i);
+
+			idx_t &ht_offset_and_salt = ht_offsets_and_salt[row_index];
+			const hash_t salt = aggr_ht_entry_t::ExtractSalt(ht_offset_and_salt);
+			idx_t ht_offset;
 
-	// use bitmask to get position in array
-	ApplyBitmask(hashes, count);
+			bool occupied;
+			bool salt_match;
 
-	hashes.Flatten(count);
-	D_ASSERT(hashes.GetVectorType() == VectorType::FLAT_VECTOR);
+			aggr_ht_entry_t entry;
 
-	auto pointers = reinterpret_cast<atomic<data_ptr_t> *>(hash_map.get());
-	auto indices = FlatVector::GetData<hash_t>(hashes);
+			// increment the ht_offset_and_salt of the entry as long as next entry is occupied and salt does not match
+			while (true) {
+				ht_offset = ht_offset_and_salt & aggr_ht_entry_t::POINTER_MASK;
+				atomic<aggr_ht_entry_t> &atomic_entry = entries[ht_offset];
+				entry = atomic_entry.load(std::memory_order_relaxed);
+				occupied = entry.IsOccupied();
+				salt_match = entry.GetSalt() == salt;
+
+				// condition for incrementing the ht_offset: occupied and row_salt does not match -> move to next entry
+				if (!occupied || salt_match) {
+					break;
+				}
+
+				IncrementAndWrap(ht_offset_and_salt, capacity_mask);
+			}
+
+			if (!occupied) { // insert into free
+				atomic<aggr_ht_entry_t> &atomic_entry = entries[ht_offset];
+				data_ptr_t row_ptr = row_ptrs_to_insert[row_index];
+				data_ptr_t potential_collided_ptr =
+				    InsertRowToEntry<PARALLEL, true>(atomic_entry, row_ptr, salt, ht->pointer_offset);
+
+				if (PARALLEL) {
+					// if the insertion was not successful, the entry was occupied in the meantime, so we have to
+					// compare the keys and insert the row to the next entry
+					if (potential_collided_ptr) {
+						// if the entry was occupied, we need to compare the keys and insert the row to the next entry
+						// we need to compare the keys and insert the row to the next entry
+						state.salt_match_sel.set_index(salt_match_count, row_index);
+						row_ptr_insert_to[salt_match_count] = potential_collided_ptr;
+						salt_match_count += 1;
+					}
+				}
+
+			} else { // compare with full
+				state.salt_match_sel.set_index(salt_match_count, row_index);
+				row_ptr_insert_to[salt_match_count] = entry.GetPointer();
+				salt_match_count += 1;
+			}
+		}
+
+		// at this step, for all the pointers_result_v we stepped either until we found an empty entry or an entry with
+		// a matching salt, we need to compare the keys for the ones that have a matching salt
+		if (salt_match_count != 0) {
```

Comment:
Can we move this part of the code that scans/compares to ensure chains have the same keys to a separate function with a descriptive name? Then this ~150 line function can be split into two functions, making the code more organized.